libs/community/langchain_community/chat_models/githubllm.py

```diff
@@ -0,0 +1,314 @@
+import os
+import logging
+import time
+from typing import Any, Dict, List, Mapping, Optional, Iterator, Union, AsyncIterator
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+
+import requests
+from langchain_core.callbacks import CallbackManagerForLLMRun, AsyncCallbackManagerForLLMRun
+from langchain_core.language_models.chat_models import BaseChatModel
+from langchain_core.messages import BaseMessage, AIMessage, HumanMessage, SystemMessage
+from langchain_core.outputs import ChatResult, ChatGeneration, ChatGenerationChunk
+from langchain_core.pydantic_v1 import Extra, Field, root_validator, SecretStr
+
+logger = logging.getLogger(__name__)
+
+def run_in_executor(executor, func, *args, **kwargs):
+    loop = asyncio.get_event_loop()
+    return loop.run_in_executor(executor, func, *args, **kwargs)
+
+def _convert_message_to_dict(message: BaseMessage) -> dict:
+    if isinstance(message, SystemMessage):
+        return {"role": "system", "content": message.content}
+    elif isinstance(message, HumanMessage):
+        return {"role": "user", "content": message.content}
+    elif isinstance(message, AIMessage):
+        return {"role": "assistant", "content": message.content}
+    else:
+        raise ValueError(f"Got unknown type {message}")
+
+def _convert_dict_to_message(message_dict: Mapping[str, Any]) -> BaseMessage:
+    role = message_dict["role"]
+    content = message_dict.get("content", "")
+    if role == "user":
+        return HumanMessage(content=content)
+    elif role == "assistant":
+        return AIMessage(content=content)
+    elif role == "system":
+        return SystemMessage(content=content)
+    else:
+        raise ValueError(f"Got unknown role {role}")
+
+class GithubLLM(BaseChatModel):
+    """GitHub LLM with Azure Fallback"""
+
+    github_endpoint_url: str = "https://models.inference.ai.azure.com/chat/completions"
+    model: str
+    github_api_key: SecretStr = Field(default_factory=lambda: SecretStr(os.environ.get("GITHUB_TOKEN", "")))
+    azure_api_key: SecretStr = Field(default_factory=lambda: SecretStr(os.environ.get("AZURE_API_KEY", "")))
```

Comment:
does this just change the auth header? would anyone ever want to instantiate with both of these keys? should we just accept `api_key`, let the user decide what key to provide, and set based on env vars in `validate_environment`?

Reply:
well, my thinking is that using the github PAT is completely free for a while , and if they are both instantiated then there is a "graceful fallback" to the azure account